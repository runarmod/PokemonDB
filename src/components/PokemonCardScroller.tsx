import DownArrow from "../assets/down_arrow.svg";
import UpArrow from "../assets/up_arrow.svg";
import "../styles/PokemonCardScroller.css";
import { useAppContext } from "../utils";

const PokemonCardScroller = ({ children }: { children: React.ReactNode }) => {
    const { selectedPokemonId, changeSelectedPokemonId, currentPokemonList } =
        useAppContext();

    function handleUpClick(): void {
        const index = currentPokemonList
            .map((pokemon) => pokemon.id)
            .indexOf(selectedPokemonId);
        if (index <= 0) {
            changeSelectedPokemonId(
                currentPokemonList[currentPokemonList.length - 1].id
            );
        } else {
            changeSelectedPokemonId(currentPokemonList[index - 1].id);
        }
    }

    function handleDownClick(): void {
        const index = currentPokemonList
            .map((pokemon) => pokemon.id)
            .indexOf(selectedPokemonId);
        if (index === currentPokemonList.length - 1) {
            changeSelectedPokemonId(currentPokemonList[0].id);
        } else {
            changeSelectedPokemonId(currentPokemonList[index + 1].id);
        }
    }

    return (
        <div id="ScrollerWrapper">
            {currentPokemonList.length > 1 && (
                <button onClick={handleUpClick}>
                    <img src={UpArrow} alt="Up Arrow" />
                </button>
            )}
            {children}
            {currentPokemonList.length > 1 && (
                <button onClick={handleDownClick}>
                    <img src={DownArrow} alt="Down Arrow" />
                </button>
            )}
        </div>
    );
};

export default PokemonCardScroller;
