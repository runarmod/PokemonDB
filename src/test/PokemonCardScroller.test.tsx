import "@testing-library/jest-dom";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, Mock, vi } from "vitest";
import PokemonCardScroller from "../components/PokemonCardScroller";
import { useAppContext } from "../utils";
import mockPokemonData from "./mocks/mockPokemonData.json";

vi.mock(import("../utils"), async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        useAppContext: vi.fn(),
    };
});

// Note: Uses div as child, as we don't want to test the PokemonCard component here
describe("PokemonCardScroller - General tests", () => {
    it("should render children", () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            changeSelectedPokemonId: vi.fn(),
            currentPokemonList: [],
        });
        render(
            <PokemonCardScroller>
                <div>PokemonCard</div>
            </PokemonCardScroller>
        );
        expect(screen.getByText("PokemonCard")).toBeInTheDocument();
    });

    it("should not render arrow buttons when pokemon list is empty", () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            changeSelectedPokemonId: vi.fn(),
            currentPokemonList: [],
        });
        render(
            <PokemonCardScroller>
                <div>PokemonCard</div>
            </PokemonCardScroller>
        );

        expect(screen.queryByAltText("Up Arrow")).not.toBeInTheDocument();
        expect(screen.queryByAltText("Down Arrow")).not.toBeInTheDocument();
    });

    it("should render arrow buttons when pokemon list is not empty", () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            changeSelectedPokemonId: vi.fn(),
            currentPokemonList: mockPokemonData,
        });
        render(
            <PokemonCardScroller>
                <div>PokemonCard</div>
            </PokemonCardScroller>
        );

        expect(screen.getByAltText("Up Arrow")).toBeInTheDocument();
        expect(screen.getByAltText("Down Arrow")).toBeInTheDocument();
    });

    it("should change selectedPokemonId when clicking down arrow", async () => {
        const mockChangeSelectedPokemonId = vi.fn();
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            changeSelectedPokemonId: mockChangeSelectedPokemonId,
            currentPokemonList: mockPokemonData,
        });
        render(
            <PokemonCardScroller>
                <div>PokemonCard</div>
            </PokemonCardScroller>
        );

        const downArrow = screen.getByAltText("Down Arrow");
        expect(downArrow).toBeInTheDocument();
        userEvent.click(downArrow);

        await waitFor(() => {
            expect(mockChangeSelectedPokemonId).toHaveBeenCalledTimes(1);
            expect(mockChangeSelectedPokemonId).toHaveBeenCalledWith(2);
        });
    });

    it("should change selectedPokemonId when clicking up arrow", async () => {
        const mockChangeSelectedPokemonId = vi.fn();
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 2,
            changeSelectedPokemonId: mockChangeSelectedPokemonId,
            currentPokemonList: mockPokemonData,
        });
        render(
            <PokemonCardScroller>
                <div>PokemonCard</div>
            </PokemonCardScroller>
        );

        const upArrow = screen.getByAltText("Up Arrow");
        expect(upArrow).toBeInTheDocument();
        userEvent.click(upArrow);

        await waitFor(() => {
            expect(mockChangeSelectedPokemonId).toHaveBeenCalledTimes(1);
            expect(mockChangeSelectedPokemonId).toHaveBeenCalledWith(1);
        });
    });

    it("should loop around when clicking up arrow from first Pokemon", async () => {
        const mockChangeSelectedPokemonId = vi.fn();
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            changeSelectedPokemonId: mockChangeSelectedPokemonId,
            currentPokemonList: mockPokemonData,
        });
        render(
            <PokemonCardScroller>
                <div>PokemonCard</div>
            </PokemonCardScroller>
        );

        const upArrow = screen.getByAltText("Up Arrow");
        expect(upArrow).toBeInTheDocument();
        userEvent.click(upArrow);

        await waitFor(() => {
            expect(mockChangeSelectedPokemonId).toHaveBeenCalledTimes(1);
            expect(mockChangeSelectedPokemonId).toHaveBeenCalledWith(4);
        });
    });

    it("should loop around when clicking down arrow from last Pokemon", async () => {
        const mockChangeSelectedPokemonId = vi.fn();
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 4,
            changeSelectedPokemonId: mockChangeSelectedPokemonId,
            currentPokemonList: mockPokemonData,
        });
        render(
            <PokemonCardScroller>
                <div>PokemonCard</div>
            </PokemonCardScroller>
        );

        const downArrow = screen.getByAltText("Down Arrow");
        expect(downArrow).toBeInTheDocument();
        userEvent.click(downArrow);

        await waitFor(() => {
            expect(mockChangeSelectedPokemonId).toHaveBeenCalledTimes(1);
            expect(mockChangeSelectedPokemonId).toHaveBeenCalledWith(1);
        });
    });
});

describe("PokemonCardScroller - Snapshot test", () => {
    it("matches snapshot", () => {
        (useAppContext as Mock).mockReturnValue({
            selectedPokemonId: 1,
            changeSelectedPokemonId: vi.fn(),
            currentPokemonList: mockPokemonData,
        });

        const { asFragment } = render(
            <PokemonCardScroller>
                <div>PokemonCard</div>
            </PokemonCardScroller>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
