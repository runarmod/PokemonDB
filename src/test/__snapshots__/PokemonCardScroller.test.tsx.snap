// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonCardScroller - Snapshot test > matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="ScrollerWrapper"
  >
    <button>
      <img
        alt="Up Arrow"
        src="/src/assets/up_arrow.svg"
      />
    </button>
    <div>
      PokemonCard
    </div>
    <button>
      <img
        alt="Down Arrow"
        src="/src/assets/down_arrow.svg"
      />
    </button>
  </div>
</DocumentFragment>
`;
